File: anagrams_scratch.py
def anagrams(string1, string2):
    string1 = string1.replace(' ', '').lower()
    string2 = string2.replace(' ', '').lower()

    if len(string1) != len(string2):
        return False

    countFlag = True

    charFreqs1 = [0] * 26
    charFreqs2 = [0] * 26

    for i in range(len(string1)):
        # ord() will get the ASCII value of the character
        charFreqs1[ord(string1[i]) - ord('a')] += 1
        charFreqs2[ord(string2[i]) - ord('a')] += 1

    for i in range(26):
        if charFreqs1[i] != charFreqs2[i]:
            countFlag = False

    return countFlag

# testing
print("Clint Eastwood and Old West Action: " + str(anagrams("Clint Eastwood", "Old West Action")))
print("Dog and God: " + str(anagrams("dog", "God")))
print("Racecar and racecars: " + str(anagrams("racecar", "racecars")))
print("Scars and racks: " + str(anagrams("scars", "racks")))

File: anagrams_scratch.java
public class anagrams_scratch
{
    public static void main(String[] args)
    {
        System.out.println("Clint Eastwood and Old West Action: " + anagrams("Clint Eastwood", "Old West Action"));
        System.out.println("Dog and God: " + anagrams("dog", "God"));
        System.out.println("Public relations and crap built on lies: " + anagrams("Public relations", "crap built on lies"));
        System.out.println("Racecar and racecars: " + anagrams("Racecar", "racecars"));
        System.out.println("Cats and scat: " + anagrams("Cats", "scat"));
        System.out.println("Scars and racks: " + anagrams("scars", "racks"));
    }

    public static boolean anagrams(String string1, String string2)
    {
        // count letter frequencies of both strings, then match based on that
        string1 = string1.toLowerCase().replace(" ", "");
        string2 = string2.toLowerCase().replace(" ", "");

        if(string1.length() != string2.length())
        {
            return false;
        }

        boolean countFlag = true;
        int[] charFreqs1 = new int[26];
        int[] charFreqs2 = new int[26];
        
        for(int i = 0; i < string1.length(); i++)
        {
            charFreqs1[string1.charAt(i) - 'a'] += 1;
            charFreqs2[string2.charAt(i) - 'a'] += 1;
        }

        // check if the frequencies match
        for(int i = 0; i < 26; i++) {
            if(charFreqs1[i] != charFreqs2[i])
            {
                countFlag = false;
            }
        }

        return countFlag;
    }
}